lib/m2ts/caption-stream.js

```diff
@@ -727,7 +735,7 @@ Cea608Stream.prototype.setRollUp = function(pts, newBaseRow) {
 
   if (newBaseRow !== undefined && newBaseRow !== this.row_) {
     // move currently displayed captions (up or down) to the new base row
-    for (var i = 0; i < this.rollUpRows_; i++) {
+    for (var i = 0; i < this.rollUpRows_ && this.row_ - i >= 0; i++) {
```

Comment:
Can you explain this a little more?

Reply:
I think maybe if you move the bounds-checking block just below to above this, you won't need this check and the row shifting will happen correctly...?

Reply:
I suppose it depends on how we want to display the data. If the incorrect data was `baseRow = 1; rollUpRows = 3`, would we expect to send the following as the rollup captions:

```
top row
declared base row

```

or 

```

some content
corrected base row
```

Reply:
Going to go with the following, since the spec recommends defaulting to the number of rows, which I think would imply that the decoder should behave as though we're in roll-up 3 rows and we'll get data in a way that makes sense for 3 rows.

```

some content
corrected base row
```

Reply:
Ok, so consider this. Our `this.row_` is 12, `this.rollUpRows_` is 3, and we've got the caption data

```
10 Some text
11 Some other text
12 More text
```
with rows 0-9 and 13-14 being `''`. Then we get a PAC which switches the base row to row 0. The iteration of this block would look like this:
```
i = 0; i < this.rollUpRows_ && this.row_ - i >= 0 === true
this.displayed[0-0] = this.displayed[12-0];
this.displayed[12-0] = '';

i = 1; i < this.rollUpRows_ && this.row_ - i >= 0 === true
this.displayed[0-1] = this.displayed[12-1]; // sad panda
this.displayed[12-1] = '';

i = 2; i < this.rollUpRows_ && this.row_ - i >= 0 === true
this.displayed[0-2] = this.displayed[12-2]; // even sadder panda
this.displayed[12-2] = '';
```

`this.displayed_` now looks like this:

```
-2 Some text
-1 Some other text
 0 More text
```
with rows 1-14 being `''`.

I think you meant to use `newBaseRow` instead of `this.row_` but that would still result in a problem:
```
i = 0; i < this.rollUpRows_ && newBaseRow - i >= 0 === true
this.displayed[0-0] = this.displayed[12-0];
this.displayed[12-0] = '';

i = 1; i < this.rollUpRows_ && newBaseRow - i >= 0 === false
i = 2; i < this.rollUpRows_ && newBaseRow - i >= 0 === false
```

as this gives the caption data

```
 0 More text
10 Some text
11 Some other text
```

(Note in particular how it's out of order now.) If instead of the logic below this block, you have just above this block the following:

```javascript
if (newBaseRow - this.rollUpRows_ + 1 < 0) {
  newBaseRow = this.rollUpRows - 1;
}
```

then the iterations are as follows, with no additional clause needed in the `for`
```
// newBaseRow is now 2 
i = 0; i < this.rollUpRows_ === true
this.displayed[2-0] = this.displayed[12-0];
this.displayed[12-0] = '';

i = 1; i < this.rollUpRows_ === true
this.displayed[2-1] = this.displayed[12-1];
this.displayed[12-1] = '';

i = 2; i < this.rollUpRows_ === true
this.displayed[2-2] = this.displayed[12-2];
this.displayed[12-2] = '';
// joyous panda
```

which gives a `this.displayed_` of the desired

```
 0 Some text
 1 Some other text
 2 More text
```

and rows 3-14 being `''`. All row data are valid (i.e. they're in rows 0-14), and we've properly kept all rows and they're in the right order.

Now, we actually probably want to do this outside of `setRollUp`, since now you've got conflicting info between `this.row_,` `this.topRow_`, and `this.rollUpRows_`. Like say, [inside the `if` block here](https://github.com/videojs/mux.js/blob/3e185285822d248779cc661de012c0211d7a7cd3/lib/m2ts/caption-stream.js#L464), before the call to `setRollUp`, like so:

```javascript
if (this.mode_ === 'rollUp') {
  if (row - this.rollUpRows_ + 1 < 0) {
    row = this.rollUpRows - 1;
  }
  this.setRollUp(packet.pts, row);
}
```

Then the following `if (row !== this.row_)` won't fire every time we get a PAC (which I've seen happen after every `RU2`/`RU3` command), and we wouldn't need any code changes within `setRollUp`. I'm pretty sure everything else will play nicely, since we're effectively changing the PAC from an invalid one to a valid one, before it get processed by anything.

Phew, sorry for the wall of text.

Reply:
I don't think it's necessary since we're dealing with a limited number of possibilities. In pop-on and paint-on, we'll only ever write to rows 0-14, since PAC row === row we write text to. In rollup, our set of possible rows to write text to is -3 -> 14 (PAC possible rows - rollUpRows). So we only need to guard against the row being < 0.